### ISSUE TITLE
Connections could be considered timed out when host time jumps forward

### ISSUE THREAD

**Issue by wonder-mice** [warmcat/libwebsockets#274]
from service.c:
```
	if (sec > wsi->pending_timeout_limit) {
		lwsl_info("TIMEDOUT WAITING on %d\n", wsi->pending_timeout);
```
Both `sec` and `pending_timeout_limit` are set using time() function. Time returned by that function can jump (because of timezone change if internal clock set to local time, user change or NTP).
That means that at any time any connection(s) could be considered timed out.

Recommendations:
a. Simplify the library and let clients to handle timeouts manually
b. Use monotonous realtime time source

**Comment by warmcat** [comment]
Timeouts are a basic feature of practical socket libraries, it does not belong in user code, being repeated in a variety of reimplementations of varying quality, or worse, missing.

You are right that pending timeouts might get applied if the time jumps forward.  But these are only active during networking activity that is anyway expected to be prone to fail, stick or get lost for a multitude of reasons.... that is why there is a timeout protecting that process.  While not engaged in some critical communication, the timeout is not pending on the connection and it makes no problem; established connections are unaffected for example.  So I think this is okay as it is.

**Comment by wonder-mice** [comment]
This statement is equivalent to "This library makes your connections *less* reliable and *more* unstable, but this is good enough for you anyway".

But it's all up to you of course.

**Comment by andrew-canaday** [comment]
We could toss together a patch that checks for the presence of
*get_clocktime* on the target system and leverages the monotonic clock
therein, if present and defaults to *time*, otherwise. Maybe I'll take a
peek at this when I get a chance.

**Comment by JoakimSoderberg** [comment]
There is a good monotonic clock implementation in:
https://github.com/libevent/libevent

Supports Linux, OSX and Windows

**Comment by andrew-canaday** [comment]
I'm not opposed! However, it does increase external dependencies. For
libev, the loop timer is also guaranteed to be monotonic - though we are
not currently leveraging this for timeouts in ev builds. As an additional
option - again, just for libev builds - the Times is open sourcing our
internal timeout management library (libevtom) sometime in the next week or
two, that'd also be an option for ev builds.

Having said that, it'd also be nice to have neat solution that worked for
all build types and didn't require an additional library - though I'm not
sure how practical that is.

**Comment by andrew-canaday** [comment]
Reonsidering, evtom may not be practical. I have to review current timeouts
in libwebsockets, but I think the usage for libev and non-libev servers
would likely be too different if it were utilized. Probably best route is
libevent, internal, or leave it for now.

On Mon, Apr 27, 2015 at 5:16 PM, Canaday, Andrew <andrew.canaday@nytimes.com
> wrote:

> I'm not opposed! However, it does increase external dependencies. For
> libev, the loop timer is also guaranteed to be monotonic - though we are
> not currently leveraging this for timeouts in ev builds. As an additional
> option - again, just for libev builds - the Times is open sourcing our
> internal timeout management library (libevtom) sometime in the next week or
> two, that'd also be an option for ev builds.
>
> Having said that, it'd also be nice to have neat solution that worked for
> all build types and didn't require an additional library - though I'm not
> sure how practical that is.
>
> On Mon, Apr 27, 2015 at 5:12 PM, Joakim Söderberg <

**Comment by andrew-canaday** [comment]
Sorry for the chain-emailing, here. Another thought, libevent also comes
with a significant amount of baggage, no? A built in HTTP server + DNS
resolution facilities? Not trying to shoot your idea down or be contrarian
- this is another area in which I really don't have much stake in the game;
linking against libevent is *also* an easy change for us. However, it might
be kind of clunky for other users, unless you can conditionally link only
some facilities (I haven't used libevent - so I apologize for my ignorance).

On Mon, Apr 27, 2015 at 5:24 PM, Canaday, Andrew <andrew.canaday@nytimes.com
> wrote:

> Reonsidering, evtom may not be practical. I have to review current
> timeouts in libwebsockets, but I think the usage for libev and non-libev
> servers would likely be too different if it were utilized. Probably best
> route is libevent, internal, or leave it for now.
>
> On Mon, Apr 27, 2015 at 5:16 PM, Canaday, Andrew <

**Comment by wonder-mice** [comment]
Since I started all this...

A function that returns monotonous time could be pretty simple, you don't need huge and bloated libevent for that. It could be as simple as that:

```c
	struct rt_ctx
	{
	#ifdef __linux__
		bool monotonic;
	#elif defined(__APPLE__) && defined(__MACH__)
		struct mach_timebase_info timebase;
	#endif
	};

	void rt_init(rt_ctx *const ctx)
	{
	#if defined(__linux__)
		struct timespec ts;
		ctx->monotonic = 0 == clock_gettime(CLOCK_MONOTONIC, &ts);
	#elif defined(__APPLE__) && defined(__MACH__)
		mach_timebase_info(&ctx->timebase);
	#else
		(void)ctx;
	#endif
	}

	unsigned rt_time_sec(rt_ctx *const ctx)
	{
	#if defined(__linux__)
		if (ctx->monotonic) {
			struct timespec ts;
			clock_gettime(CLOCK_MONOTONIC, &ts);
			return ts.tv_sec;
		}
		time_t now;
		time(&now);
		return now;
	#elif defined(__APPLE__) && defined(__MACH__)
		const uint64_t t = mach_absolute_time();
		return t  * ctx->timebase.numer / ctx->timebase.denom / UNITS;
	#elif defined(_WIN32)
		return GetTickCount() / 1000;
	#else
		time_t now;
		time(&now);
		return now;
	#endif
	}
```

**Comment by warmcat** [comment]
Well, having the timeouts in lws dramatically increases consistency and reliability compared to evicting them to user code, which was the number 1 suggestion by the OP.  So we won't be doing that.

Trying to avoid pending timeout firing on clock jump is not worth creating a dependency on a second event library either, although I appreciate knowing about it.

This is simply not a practical problem, especially since ntpd jumps are usually measured in us.

But if someone thinks this is something important, how about try count the timeout checks in the wsi and confirm we sat through n checks at >= 1s intervals as well as the time test?  It will make timeouts more generous under load but otherwise not too objectionable.  And the cost is almost nothing.

**Comment by andrew-canaday** [comment]
> Well, having the timeouts in lws dramatically increases consistency and
> reliability compared to evicting them to user code, which was the number 1
> suggestion by the OP. So we won't be doing that.
>
Okay, makes sense.


> Trying to avoid pending timeout firing on clock jump is not worth creating
> a dependency on a second event library either, although I appreciate
> knowing about it.
>
Okay, makes sense.

> This is simply not a practical problem, especially since ntpd jumps are
> usually measured in us.
>
> But if someone thinks this is something important, how about try count the
> timeout checks in the wsi and confirm we sat through n checks at >= 1s
> intervals as well as the time test? It will make timeouts more generous
> under load but otherwise not too objectionable. And the cost is almost
> nothing.
>
Seems workable. :)

Ciao!

**Comment by warmcat** [comment]
@wonder-mice what is the license of that code?

**Comment by JoakimSoderberg** [comment]
@andrew-canaday  Oh you misunderstood me, I did not mean link with libevent. I meant borrow the code for the monotonic timer ONLY from that code base. It's very easy to rip out just that part, and the implementation is nice. But the above @wonder-mice solution seems nice and compact.

**Comment by wonder-mice** [comment]
@warmcat Are you kidding me? :) But since you asked - [The Use It Well License](https://github.com/wonder-mice/zf_queue/blob/master/LICENSE). Keep in mind that it's just a sketch, not a fully functional code.

**Comment by JoakimSoderberg** [comment]
Here is the libevent implementation:
https://github.com/libevent/libevent/blob/master/evutil_time.c#L148

It takes care of a lot of special cases on different platforms and what not. And adjusts the clock to always guarantee that whatever monotonic source it gets is really monotonic.

On some systems the clock stops on sleeps and such...

**Comment by andrew-canaday** [comment]
Right on, and right on. Yeah, I like it. Nice and clean. Apple warns that
using the data from *mach_timebase_info* in constructions like
"ctx->timebase.numer / ctx->timebase.denom / UNITS" may suffer arithmetic
overflow here
<https://developer.apple.com/library/mac/qa/qa1398/_index.html>, under
listing 2. I don't know how big of a concern it is. This, again, is not a
huge concern to me (we're Linux-only), but thought I'd point it out.

**Comment by warmcat** [comment]
Guys... I don't think we ever had such a long discussion about something so small.

Taking from libevent would be OK but the licence requires their "copyright message in the binary" whatever that means.  And it wasn't clear from looking at the code how much trailing libevent guts will come out with the bit we want.

I dunno where wm's code teleported in from and there is no license or claim of authorship attached; clearly it came from an existing proejct of some kind.

If somebody cares enough about this non-problem to send a patch, okay, otherwise we're just trying to fix a non-problem by talking about it.

**Comment by wonder-mice** [comment]
@andrew-canaday It's a valid concern. I mentioned in my previous message that "it's just a sketch, not a fully functional code".

There are several ways to address that issue. The simplest way is just to precompute `uint64_t ratio = UNITS * timebase.denom / timebase.numer` and the time in seconds will be `t / ratio`. It's good enough from practical standpoint for 1sec precision.

@warmcat

> I dunno where wm's code teleported in from...

Common, I just crafted it in TextEdit exclusively for that message thread.
